In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\users\rendo\08-2020 Instacart Basket Analysis'

In [3]:
path

'C:\\users\\rendo\\08-2020 Instacart Basket Analysis'

In [5]:
df_orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [6]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  price_range_loc  \
0         Soda        77              7     9.0   both              NaN   
1         Soda        77              7     9.0   both              NaN   
2         Soda        77              7     9.0   both              NaN   
3         Soda        77              7     9.0   both              NaN   
4         Soda        77              7     9.0   both              NaN   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy   Regularly busy        Average orders  
1  Mid-range product  Regularly busy  Least busy days        Average orders  
2  Mid-range product  Regularly busy  Least busy days           Most orders  
3  Mid-range product      Least busy  Least busy days        Average orders  
4  Mid-range product      Least busy  Least busy days           Most orders

In [7]:
df_orders_products_merged.shape

(32404859, 19)

In [8]:
# Creating subset

df = df_orders_products_merged[:1000000]

In [9]:
df.shape

(1000000, 19)

In [10]:
df.head(10)

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   
5   3367565        1             6                   2            7   
6    550135        1             7                   1            9   
7   3108588        1             8                   1           14   
8   2295261        1             9                   1           16   
9   2550362        1            10                   4            8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  price_range_loc  \
0         Soda        77              7     9.0   both              NaN   
1         Soda        77              7     9.0   both              NaN   
2         Soda        77              7     9.0   both              NaN   
3         Soda        77              7     9.0   both              NaN   
4         Soda        77              7     9.0   both              NaN   
5         Soda        77              7     9.0   both              NaN   
6         Soda        77              7     9.0   both              NaN   
7         Soda        77              7     9.0   both              NaN   
8         Soda        77              7     9.0   both              NaN   
9         Soda        77              7     9.0   both              NaN   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy   Regularly busy        Average orders  
1  Mid-range product  Regularly busy  Least busy days        Average orders  
2  Mid-range product  Regularly busy  Least busy days           Most orders  
3  Mid-range product      Least busy  Least busy days        Average orders  
4  Mid-range product      Least busy  Least busy days           Most orders  
5  Mid-range product  Regularly busy   Regularly busy        Average orders  
6  Mid-range product  Regularly busy     Busiest days           Most orders  
7  Mid-range product  Regularly busy     Busiest days           Most orders  
8  Mid-range product  Regularly busy     Busiest days           Most orders  
9  Mid-range product      Least busy  Least busy days        Average orders

In [11]:
df.groupby('product_name')

### aggregate data

In [12]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [13]:
# Different way of finding mean of 'order_number'

df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [14]:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

### aggregate with transform

In [15]:
df_orders_products_merged['max_order'] = df_orders_products_merged.groupby('user_id')['order_number'].transform(np.max)

In [16]:
df_orders_products_merged.head(25)

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0    2539329        1             1                   2            8   
1    2398795        1             2                   3            7   
2     473747        1             3                   3           12   
3    2254736        1             4                   4            7   
4     431534        1             5                   4           15   
5    3367565        1             6                   2            7   
6     550135        1             7                   1            9   
7    3108588        1             8                   1           14   
8    2295261        1             9                   1           16   
9    2550362        1            10                   4            8   
10   2968173       15            15                   1            9   
11   1870022       15            17                   2           16   
12   1911383       15            18                   2           11   
13   2715276       15            21                   1            9   
14    487368       15            22                   1           10   
15   2293453       19             2                   5           14   
16   1973799       19             5                   6           12   
17    532817       19             7                   4           17   
18   1573906       21            10                   3           10   
19   1593000       31            10                   3            8   
20   2231262       31            17                   3           11   
21   2580647       43             6                   4           16   
22   2187180       43             9                   4           12   
23   2497897       52             1                   1            9   
24   1318871       52             2                   1           10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   
20                     8.0         196                 14          1   
21                     4.0         196                  9          0   
22                     3.0         196                  1          1   
23                     NaN         196                  4          0   
24                     7.0         196                  5          1   

   product_name  aisle_id  department_id  prices _merge  price_range_loc  \
0          Soda        77              7     9.0   both              NaN   
1          Soda        77              7     9.0   both              NaN   
2          Soda        77     

In [17]:
df_orders_products_merged.loc[df_orders_products_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [18]:
df_orders_products_merged.loc[(df_orders_products_merged['max_order'] <= 40) & (df_orders_products_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [19]:
df_orders_products_merged.loc[df_orders_products_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [20]:
df_orders_products_merged['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [21]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge  price_range_loc  \
0         Soda  ...              7     9.0    both              NaN   
1         Soda  ...              7     9.0    both              NaN   
2         Soda  ...              7     9.0    both              NaN   
3         Soda  ...              7     9.0    both              NaN   
4         Soda  ...              7     9.0    both              NaN   

     price_range_loc     busiest_day     busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy   Regularly busy        Average orders   
1  Mid-range product  Regularly busy  Least busy days        Average orders   
2  Mid-range product  Regularly busy  Least busy days           Most orders   
3  Mid-range product      Least busy  Least busy days        Average orders   
4  Mid-range product      Least busy  Least busy days           Most orders   

  max_order  loyalty_flag  
0        10  New customer  
1        10  New customer  
2        10  New customer  
3        10  New customer  
4        10  New customer  

[5 rows x 21 columns]

In [22]:
df_orders_products_merged[['user_id', 'max_order', 'loyalty_flag']].sample(30)

user_id  max_order      loyalty_flag
9684971    200820         28  Regular customer
829968     160287         26  Regular customer
22299978   173006         28  Regular customer
1177198     66399         15  Regular customer
3748917    146012         31  Regular customer
30038098   197591         40  Regular customer
7364572     70509         33  Regular customer
26776576     2754         15  Regular customer
11152500   129182         71    Loyal customer
6383370     93922         29  Regular customer
8485428    194280         33  Regular customer
9015101    188539         62    Loyal customer
20791650   125158         99    Loyal customer
12546332    15556         17  Regular customer
23123243   124949         82    Loyal customer
28644070     3974         48    Loyal customer
26912982    25549         29  Regular customer
5738607    145936         28  Regular customer
7387304    163993         42    Loyal customer
24678461   115423         99    Loyal customer
3536418     97349         20  Regular customer
748265      65256          7      New customer
28278679   166182         64    Loyal customer
7628204     72776          9      New customer
13190242   167132         19  Regular customer
25388753     6179          5      New customer
18635782   102303          9      New customer
3325428     56826         19  Regular customer
8903310    173974          4      New customer
9829465     63020          6      New customer

# 4.8 task

In [23]:
# step 1 - create new notebook
# step 2
df_orders_products_merged.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [24]:
# step 3 
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

#### average of order_number from subset is greater than entire dateframe except department_id=17

### step 4
##### see output 22

In [25]:
# step 5
print('The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:')
df_orders_products_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

#### new customers have a higher average purchase than loyal and regular customers. loyal customers also spend less than regular customers.
##### the max value being 99999 would throw off distribution

In [26]:
# Run descriptive statistics on the 'prices' column

print('Descriptive statistics of the prices column:')
df_orders_products_merged['prices'].describe()

Descriptive statistics of the prices column:


count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

### average price is 11.9 while max is 99999. that leaves a huge gap in between data.
#### std is 495.655 which shows variation in data.

In [27]:
print('Check for abnormal values in the prices column:')
df_orders_products_merged['prices'].sort_values(ascending = False)

Check for abnormal values in the prices column:


29165979    99999.0
29165985    99999.0
29165519    99999.0
29165518    99999.0
29165517    99999.0
             ...   
7870751         1.0
7870750         1.0
7870749         1.0
7870748         1.0
30034999        1.0
Name: prices, Length: 32404859, dtype: float64

In [28]:
print('Check for abnormal values in the prices column:')

df_orders_products_merged['prices'].drop_duplicates().sort_values(ascending = False)

Check for abnormal values in the prices column:


29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

#### 99999 and 14900 prices are drastically different than the others.

In [29]:
max_price = df_orders_products_merged.loc[df_orders_products_merged['prices'].isin([99999, 14900])]

In [30]:
print('Sample output of df_orders_products_merged when the value in the prices column is either 99999 or 14900:')
max_price.sample(5)

Sample output of df_orders_products_merged when the value in the prices column is either 99999 or 14900:


order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
10032739   2439429   111178            14                   4           14   
29165861    195004   114060            54                   1            9   
10032449   2863258   100969            93                   0           20   
29166029    559592   155790             8                   2           13   
10032921   2254007   119610            19                   0            8   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
10032739                    30.0       21553                  6          1   
29165861                     4.0       33664                  7          1   
10032449                     3.0       21553                  3          1   
29166029                     8.0       33664                 12          1   
10032921                     5.0       21553                  1          1   

                              product_name  ...  department_id   prices  \
10032739  Lowfat 2% Milkfat Cottage Cheese  ...             16  14900.0   
29165861             2 % Reduced Fat  Milk  ...             16  99999.0   
10032449  Lowfat 2% Milkfat Cottage Cheese  ...             16  14900.0   
29166029             2 % Reduced Fat  Milk  ...             16  99999.0   
10032921  Lowfat 2% Milkfat Cottage Cheese  ...             16  14900.0   

          _merge     price_range_loc price_range_loc     busiest_day  \
10032739    both  High-range product             NaN      Least busy   
29165861    both  High-range product             NaN  Regularly busy   
10032449    both  High-range product             NaN     Busiest day   
29166029    both  High-range product             NaN  Regularly busy   
10032921    both  High-range product             NaN     Busiest day   

             busiest_days busiest_period_of_day max_order      loyalty_flag  
10032739  Least busy days           Most orders        14  Regular customer  
29165861     Busiest days           Most orders        78    Loyal customer  
10032449     Busiest days        Average orders        99    Loyal customer  
29166029   Regularly busy           Most orders        19  Regular customer  
10032921     Busiest days        Average orders        71    Loyal customer  

[5 rows x 21 columns]

In [31]:
print('Descriptive statistics on max_price:')
max_price.describe()

Descriptive statistics on max_price:


order_id        user_id  order_number  orders_day_of_week  \
count  5.127000e+03    5127.000000   5127.000000         5127.000000   
mean   1.708876e+06  106117.185684     15.302126            2.750536   
std    9.838956e+05   58594.982659     16.691849            2.010892   
min    4.300000e+01      17.000000      1.000000            0.000000   
25%    8.642320e+05   58136.000000      4.000000            1.000000   
50%    1.693799e+06  106021.000000      9.000000            3.000000   
75%    2.567244e+06  156389.000000     20.000000            5.000000   
max    3.421047e+06  206049.000000     99.000000            6.000000   

       hour_of_day  days_since_prior_order    product_id  add_to_cart_order  \
count  5127.000000             4776.000000   5127.000000        5127.000000   
mean     13.473376               11.899079  23201.815682           7.986347   
std       4.137326                9.068019   4153.742747           6.868889   
min       0.000000                0.000000  21553.000000           1.000000   
25%      10.000000                5.000000  21553.000000           3.000000   
50%      13.000000                8.000000  21553.000000           6.000000   
75%      16.000000               16.000000  21553.000000          11.000000   
max      23.000000               30.000000  33664.000000          72.000000   

         reordered     aisle_id  department_id        prices    max_order  
count  5127.000000  5127.000000         5127.0   5127.000000  5127.000000  
mean      0.613614   104.732592           16.0  26485.547494    29.791691  
std       0.486968     8.231346            0.0  29186.636448    24.929207  
min       0.000000    84.000000           16.0  14900.000000     3.000000  
25%       0.000000   108.000000           16.0  14900.000000    12.000000  
50%       1.000000   108.000000           16.0  14900.000000    21.000000  
75%       1.000000   108.000000           16.0  14900.000000    41.000000  
max       1.000000   108.000000           16.0  99999.000000    99.000000

### 14900 and 99999 are from department_id=16 and product_id= 21553, 33664

In [32]:
df_orders_products_merged.loc[df_orders_products_merged['product_id'].isin([21553, 33664])]

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
10030345    912404       17            12                   2           14   
10030346    603376       17            22                   6           16   
10030347   3264360      135             2                   2           21   
10030348    892534      135             3                   0            8   
10030349    229704      342             8                   1           19   
...            ...      ...           ...                 ...          ...   
29166209   2249946   204099            29                   0            8   
29166210   2363282   204099            31                   0            9   
29166211   3181945   204395            13                   3           15   
29166212   2486215   205227             7                   3           20   
29166213   2232716   205227            10                   0           11   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
10030345                     5.0       21553                  5          0   
10030346                     4.0       21553                  3          1   
10030347                    13.0       21553                  6          0   
10030348                    12.0       21553                  3          1   
10030349                    30.0       21553                  9          0   
...                          ...         ...                ...        ...   
29166209                     4.0       33664                  1          0   
29166210                     2.0       33664                  1          1   
29166211                     8.0       33664                 25          0   
29166212                     4.0       33664                  8          0   
29166213                     4.0       33664                 23          1   

                              product_name  ...  department_id   prices  \
10030345  Lowfat 2% Milkfat Cottage Cheese  ...             16  14900.0   
10030346  Lowfat 2% Milkfat Cottage Cheese  ...             16  14900.0   
10030347  Lowfat 2% Milkfat Cottage Cheese  ...             16  14900.0   
10030348  Lowfat 2% Milkfat Cottage Cheese  ...             16  14900.0   
10030349  Lowfat 2% Milkfat Cottage Cheese  ...             16  14900.0   
...                                    ...  ...            ...      ...   
29166209             2 % Reduced Fat  Milk  ...             16  99999.0   
29166210             2 % Reduced Fat  Milk  ...             16  99999.0   
29166211             2 % Reduced Fat  Milk  ...             16  99999.0   
29166212             2 % Reduced Fat  Milk  ...             16  99999.0   
29166213             2 % Reduced Fat  Milk  ...             16  99999.0   

          _merge     price_range_loc price_range_loc     busiest_day  \
10030345    both  High-range product             NaN  Regularly busy   
10030346    both  High-range product             NaN  Regularly busy   
10030347    both  High-range product             NaN  Regularly busy   
10030348    both  High-range product             NaN     Busiest day   
10030349    both  High-range product             NaN  Regularly busy   
...          ...                 ...             ...             ...   
29166209    both  High-range product             NaN     Busiest day   
29166210    both  High-range product             NaN     Busiest day   
29166211    both  High-range product             NaN  Regularly busy   
29166212    both  High-range product             NaN  Regularly busy   
29166213    both  High-range product             NaN     Busiest day   

             busiest_days busiest_period_of_day max_order      loyalty_flag  
10030345   Regularly busy           Most orders        40  Regular customer  
10030346   Regularly busy           Most orders        40  Regular customer  
10030347   Regularly busy        Average orders         4      New customer  
10030348     Busiest days        Average orders         4 

#### prodcut_name is Lowfat 2% Milkfat Cottage Cheese and 2 % Reduced Fat Milk

In [33]:
dairy = df_orders_products_merged.loc[df_orders_products_merged['department_id'].isin([16])]
dairy_without_weird = dairy.loc[~dairy['prices'].isin([99999, 14900])]
print('Filter for records with the department_id value of 16, while excluding records without the unusually high prices:')
dairy_without_weird.sample(5)

Filter for records with the department_id value of 16, while excluding records without the unusually high prices:


order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
25570075    246856   187001            10                   2           12   
27609742   3411296    17633            13                   4           11   
5470381     570741   119534             5                   6           23   
8418135    3418827    19538            13                   4            7   
27653100   2160217   101129             5                   5           13   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
25570075                    10.0       35591                 38          0   
27609742                     5.0       48707                 10          0   
5470381                      6.0       42828                  3          1   
8418135                      7.0       44987                  2          1   
27653100                    30.0       14218                  2          1   

                                             product_name  ...  department_id  \
25570075  Natural Reduced Fat Sharp Cheddar Cheese Sticks  ...             16   
27609742                              Banana Mango Yogurt  ...             16   
5470381                              Whipped Cream Cheese  ...             16   
8418135                              Heavy Whipping Cream  ...             16   
27653100             Grade A Large Eggs Cage Free Omega 3  ...             16   

          prices  _merge  price_range_loc    price_range_loc     busiest_day  \
25570075    10.7    both              NaN  Mid-range product  Regularly busy   
27609742    10.2    both              NaN  Mid-range product      Least busy   
5470381      2.4    both              NaN  Low-range product  Regularly busy   
8418135      9.1    both              NaN  Mid-range product      Least busy   
27653100    12.8    both              NaN  Mid-range product  Regularly busy   

             busiest_days busiest_period_of_day max_order      loyalty_flag  
25570075   Regularly busy           Most orders        56    Loyal customer  
27609742  Least busy days           Most orders        27  Regular customer  
5470381    Regularly busy         Fewest orders         8      New customer  
8418135   Least busy days        Average orders        63    Loyal customer  
27653100   Regularly busy           Most orders         9      New customer  

[5 rows x 21 columns]

In [34]:
print('Maximum price in the dairy_without_weird filter:')
dairy_without_weird['prices'].max()

Maximum price in the dairy_without_weird filter:


15.0

In [35]:
reduced_fat_milk = dairy_without_weird.query('product_name.str.contains("Reduced Fat") & product_name.str.contains("2") & product_name.str.contains("Milk")')

In [36]:
print('Sample output from reduced_fat_milk:')
reduced_fat_milk.sample(10)

Sample output from reduced_fat_milk:


order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
16032959   1901583   159909            31                   4           18   
15999515   2784788    16032            34                   5           15   
20172535   2776430   194811             5                   4           14   
12865129   1679228   105772             7                   4           11   
16006159   2840002    43686            18                   2           16   
16042134   1669494   201248            81                   0           21   
16347730    388493   167945            55                   1           17   
14564741    826202   171535             4                   2           12   
16339597   2682566    62616            19                   1           18   
16012786    717904    72748             1                   2           13   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
16032959                     2.0        5785                  2          1   
15999515                     8.0        5785                  3          1   
20172535                     9.0       13166                  3          1   
12865129                    30.0        1940                  1          0   
16006159                     6.0        5785                  5          1   
16042134                     3.0        5785                  1          1   
16347730                     5.0       40174                  4          1   
14564741                    30.0       23909                  7          1   
16339597                     6.0       40174                  2          1   
16012786                     NaN        5785                  1          0   

                                  product_name  ...  department_id  prices  \
16032959           Organic Reduced Fat 2% Milk  ...             16     7.9   
15999515           Organic Reduced Fat 2% Milk  ...             16     7.9   
20172535  Organic Milk Reduced Fat, 2% Milkfat  ...             16    12.8   
12865129           Organic 2% Reduced Fat Milk  ...             16     9.1   
16006159           Organic Reduced Fat 2% Milk  ...             16     7.9   
16042134           Organic Reduced Fat 2% Milk  ...             16     7.9   
16347730           2% Reduced Fat Organic Milk  ...             16    11.4   
14564741                   2% Reduced Fat Milk  ...             16     9.2   
16339597           2% Reduced Fat Organic Milk  ...             16    11.4   
16012786           Organic Reduced Fat 2% Milk  ...             16     7.9   

          _merge  price_range_loc    price_range_loc     busiest_day  \
16032959    both              NaN  Mid-range product      Least busy   
15999515    both              NaN  Mid-range product  Regularly busy   
20172535    both              NaN  Mid-range product      Least busy   
12865129    both              NaN  Mid-range product      Least busy   
16006159    both              NaN  Mid-range product  Regularly busy   
16042134    both              NaN  Mid-range product     Busiest day   
16347730    both              NaN  Mid-range product  Regularly busy   
14564741    both              NaN  Mid-range product  Regularly busy   
16339597    both              NaN  Mid-range product  Regularly busy   
16012786    both              NaN  Mid-range product  Regularly busy   

             busiest_days busiest_period_of_day max_order      loyalty_flag  
16032959  Least busy days        Average orders        85    Loyal customer  
15999515   Regularly busy           Most orders        35  Regular customer  
20172535  Least busy days           Most orders         5      New customer  
12865129  Least busy days           Most orders        13  Regular customer  
16006159   Regularly busy           Most orders        22  Regular customer  
16042134     Busiest days        Average orders        99    Loyal customer  
16347730     Busiest days        Average orders        62    Loyal customer  
14564741   Regul

In [37]:
print('Quick statistics in reduced_fat_milk:')
reduced_fat_milk.describe()

Quick statistics in reduced_fat_milk:


order_id        user_id   order_number  orders_day_of_week  \
count  2.022380e+05  202238.000000  202238.000000       202238.000000   
mean   1.711414e+06  103040.384151      18.528684            2.748702   
std    9.876774e+05   59620.791055      18.288088            2.022770   
min    5.000000e+00      11.000000       1.000000            0.000000   
25%    8.586890e+05   51496.000000       5.000000            1.000000   
50%    1.706996e+06  102925.000000      12.000000            3.000000   
75%    2.566178e+06  154743.000000      26.000000            5.000000   
max    3.421036e+06  206209.000000      99.000000            6.000000   

         hour_of_day  days_since_prior_order     product_id  \
count  202238.000000           190556.000000  202238.000000   
mean       13.181835               10.538445   18768.761009   
std         4.193794                8.380235   13345.840890   
min         0.000000                0.000000    1940.000000   
25%        10.000000                5.000000    5785.000000   
50%        13.000000                7.000000   19820.000000   
75%        16.000000               14.000000   32478.000000   
max        23.000000               30.000000   49319.000000   

       add_to_cart_order      reordered       aisle_id  department_id  \
count      202238.000000  202238.000000  202238.000000       202238.0   
mean            5.423051       0.782692      83.716626           16.0   
std             5.397505       0.412415       6.511429            0.0   
min             1.000000       0.000000      21.000000           16.0   
25%             2.000000       1.000000      84.000000           16.0   
50%             4.000000       1.000000      84.000000           16.0   
75%             7.000000       1.000000      84.000000           16.0   
max           137.000000       1.000000      91.000000           16.0   

              prices      max_order  
count  202238.000000  202238.000000  
mean        9.086900      35.834769  
std         3.064869      26.069071  
min         1.200000       3.000000  
25%         7.900000      14.000000  
50%         9.200000      30.000000  
75%        11.400000      51.000000  
max        14.700000      99.000000

#### mean, min and max of anything with reduced fat 2 and milk are 9.08, 1.2, and 14.7. with that making more sense we can change product_id 33664 to 9.99 

In [38]:
cottage_cheese = dairy_without_weird.query('product_name.str.contains("Cottage Cheese") & product_name.str.contains("2") & product_name.str.contains("Milkfat")')

In [39]:
print('Sample output from cottage_cheese:')
cottage_cheese.sample(10)

Sample output from cottage_cheese:


order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
16192417   1960535   170497             4                   0           12   
29898873   2235826    15462             2                   1           23   
16192419   2666606   170497             6                   0           19   
16190928   1855487    44608             5                   5           16   
16191670   3349911   109952             9                   1           15   
29899170   2541808   129913             6                   5            5   
25215390   3210954   204468             2                   5           12   
16192328   3401006   162651            18                   1           18   
30232400   1972722    46506             4                   0           21   
16191190   2838038    73321             6                   2           12   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
16192417                    12.0       27790                  4          1   
29898873                    30.0        6852                  5          1   
16192419                    14.0       27790                  5          1   
16190928                     5.0       27790                  3          1   
16191670                    14.0       27790                  3          1   
29899170                    22.0        6852                 12          0   
25215390                    21.0       16257                  9          1   
16192328                     6.0       27790                  5          0   
30232400                     7.0       31030                 10          1   
16191190                    21.0       27790                  3          1   

                                               product_name  ...  \
16192417        Small Curd Lowfat 2% Milkfat Cottage Cheese  ...   
29898873         4% Milkfat Cottage Cheese Small Curd 22 OZ  ...   
16192419        Small Curd Lowfat 2% Milkfat Cottage Cheese  ...   
16190928        Small Curd Lowfat 2% Milkfat Cottage Cheese  ...   
16191670        Small Curd Lowfat 2% Milkfat Cottage Cheese  ...   
29899170         4% Milkfat Cottage Cheese Small Curd 22 OZ  ...   
25215390  2% Milkfat Low Fat Large Curd Pot Style Cottag...  ...   
16192328        Small Curd Lowfat 2% Milkfat Cottage Cheese  ...   
30232400               2% Milkfat Large Curd Cottage Cheese  ...   
16191190        Small Curd Lowfat 2% Milkfat Cottage Cheese  ...   

          department_id  prices  _merge  price_range_loc    price_range_loc  \
16192417             16     7.7    both              NaN  Mid-range product   
29898873             16     8.6    both              NaN  Mid-range product   
16192419             16     7.7    both              NaN  Mid-range product   
16190928             16     7.7    both              NaN  Mid-range product   
16191670             16     7.7    both              NaN  Mid-range product   
29899170             16     8.6    both              NaN  Mid-range product   
25215390             16     5.3    both              NaN  Mid-range product   
16192328             16     7.7    both              NaN  Mid-range product   
30232400             16     7.2    both              NaN  Mid-range product   
16191190             16     7.7    both              NaN  Mid-range product   

             busiest_day    busiest_days busiest_period_of_day max_order  \
16192417     Busiest day    Busiest days           Most orders        18   
29898873  Regularly busy    Busiest days         Fewest orders         8   
16192419     Busiest day    Busiest days        Average orders        18   
16190928  Regularly busy  Regularly busy           Most orders        43   
16191670  Regularly busy    Busiest days           Most orders        29   
29899170  Regularly busy  Regularly busy         Fewest orders        15   
25215390  Regularly busy  Regularly busy           Most orders        13   
16192328  Regularly busy    Busiest days        Average orders   

In [40]:
print('Quick statistics in cottage_cheese:')
cottage_cheese.describe()

Quick statistics in cottage_cheese:


order_id        user_id  order_number  orders_day_of_week  \
count  3.487000e+03    3487.000000   3487.000000         3487.000000   
mean   1.684106e+06  103778.688271     15.484084            2.849728   
std    9.785618e+05   58583.418992     16.808776            2.031540   
min    2.301000e+03      52.000000      1.000000            0.000000   
25%    8.438150e+05   55839.000000      4.000000            1.000000   
50%    1.673171e+06  107636.000000     10.000000            3.000000   
75%    2.535750e+06  151541.500000     20.000000            5.000000   
max    3.419042e+06  206072.000000     98.000000            6.000000   

       hour_of_day  days_since_prior_order    product_id  add_to_cart_order  \
count  3487.000000             3218.000000   3487.000000        3487.000000   
mean     13.257241               12.898073  22348.791225           5.840551   
std       4.297401                9.231128   8826.332190           5.955836   
min       0.000000                0.000000    195.000000           1.000000   
25%      10.000000                6.000000  11203.000000           2.000000   
50%      13.000000               10.000000  27790.000000           4.000000   
75%      16.000000               19.000000  27790.000000           8.000000   
max      23.000000               30.000000  31030.000000          57.000000   

         reordered     aisle_id  department_id       prices    max_order  
count  3487.000000  3487.000000         3487.0  3487.000000  3487.000000  
mean      0.697448   107.650703           16.0     7.228219    30.531116  
std       0.459429     5.502322            0.0     1.757900    25.413902  
min       0.000000    21.000000           16.0     2.400000     3.000000  
25%       0.000000   108.000000           16.0     7.700000    11.000000  
50%       1.000000   108.000000           16.0     7.700000    22.000000  
75%       1.000000   108.000000           16.0     7.700000    43.000000  
max       1.000000   108.000000           16.0    11.800000    99.000000

##### mean, min, and max for things containing cottage cheese, 2 and milkfat are 7.23, 2.4, and 11.8. with that making more sense we can change product_id 21553 to reflect the mean value.

In [41]:
df_orders_products_merged['prices'] = df_orders_products_merged['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})

In [42]:
print('The statistics of the new prices column in df_orders_products_merged:')
df_orders_products_merged['prices'].describe()

The statistics of the new prices column in df_orders_products_merged:


count    3.240486e+07
mean     7.790965e+00
std      4.241491e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

In [43]:
print('The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:')
df_orders_products_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:


prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773526  1.0  25.0
New customer      7.801190  1.0  25.0
Regular customer  7.798239  1.0  25.0

### new customer now holds highest average purchases. max value is more realistic.

In [44]:
df_orders_products_merged['average_price'] = df_orders_products_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [45]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices  _merge   price_range_loc    price_range_loc  \
0         Soda  ...     9.0    both               NaN  Mid-range product   
1         Soda  ...     9.0    both               NaN  Mid-range product   
2         Soda  ...     9.0    both               NaN  Mid-range product   
3         Soda  ...     9.0    both               NaN  Mid-range product   
4         Soda  ...     9.0    both               NaN  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day max_order  \
0  Regularly busy   Regularly busy        Average orders        10   
1  Regularly busy  Least busy days        Average orders        10   
2  Regularly busy  Least busy days           Most orders        10   
3      Least busy  Least busy days        Average orders        10   
4      Least busy  Least busy days           Most orders        10   

   loyalty_flag  average_price  
0  New customer       6.367797  
1  New customer       6.367797  
2  New customer       6.367797  
3  New customer       6.367797  
4  New customer       6.367797  

[5 rows x 22 columns]

In [46]:
df_orders_products_merged['average_price'] = df_orders_products_merged['average_price'].round(2)

In [47]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices  _merge   price_range_loc    price_range_loc  \
0         Soda  ...     9.0    both               NaN  Mid-range product   
1         Soda  ...     9.0    both               NaN  Mid-range product   
2         Soda  ...     9.0    both               NaN  Mid-range product   
3         Soda  ...     9.0    both               NaN  Mid-range product   
4         Soda  ...     9.0    both               NaN  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day max_order  \
0  Regularly busy   Regularly busy        Average orders        10   
1  Regularly busy  Least busy days        Average orders        10   
2  Regularly busy  Least busy days           Most orders        10   
3      Least busy  Least busy days        Average orders        10   
4      Least busy  Least busy days           Most orders        10   

   loyalty_flag  average_price  
0  New customer           6.37  
1  New customer           6.37  
2  New customer           6.37  
3  New customer           6.37  
4  New customer           6.37  

[5 rows x 22 columns]

In [48]:
df_orders_products_merged.loc[df_orders_products_merged['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [49]:
df_orders_products_merged.loc[df_orders_products_merged['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [50]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge   price_range_loc    price_range_loc  \
0         Soda  ...    both               NaN  Mid-range product   
1         Soda  ...    both               NaN  Mid-range product   
2         Soda  ...    both               NaN  Mid-range product   
3         Soda  ...    both               NaN  Mid-range product   
4         Soda  ...    both               NaN  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day max_order  \
0  Regularly busy   Regularly busy        Average orders        10   
1  Regularly busy  Least busy days        Average orders        10   
2  Regularly busy  Least busy days           Most orders        10   
3      Least busy  Least busy days        Average orders        10   
4      Least busy  Least busy days           Most orders        10   

   loyalty_flag average_price  spending_flag  
0  New customer          6.37    Low spender  
1  New customer          6.37    Low spender  
2  New customer          6.37    Low spender  
3  New customer          6.37    Low spender  
4  New customer          6.37    Low spender  

[5 rows x 23 columns]

In [51]:
df_orders_products_merged['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32284501
High spender      120358
Name: count, dtype: int64

In [52]:
df_orders_products_merged['median_prior_orders'] = df_orders_products_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [53]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...   price_range_loc    price_range_loc     busiest_day  \
0         Soda  ...               NaN  Mid-range product  Regularly busy   
1         Soda  ...               NaN  Mid-range product  Regularly busy   
2         Soda  ...               NaN  Mid-range product  Regularly busy   
3         Soda  ...               NaN  Mid-range product      Least busy   
4         Soda  ...               NaN  Mid-range product      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy        Average orders        10  New customer   
1  Least busy days        Average orders        10  New customer   
2  Least busy days           Most orders        10  New customer   
3  Least busy days        Average orders        10  New customer   
4  Least busy days           Most orders        10  New customer   

  average_price spending_flag  median_prior_orders  
0          6.37   Low spender                 20.5  
1          6.37   Low spender                 20.5  
2          6.37   Low spender                 20.5  
3          6.37   Low spender                 20.5  
4          6.37   Low spender                 20.5  

[5 rows x 24 columns]

In [54]:
df_orders_products_merged.loc[df_orders_products_merged['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [55]:
df_orders_products_merged.loc[(df_orders_products_merged['median_prior_orders'] > 10) & (df_orders_products_merged['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [56]:
df_orders_products_merged.loc[df_orders_products_merged['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [57]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0   2539329        1             1                   2            8   
1   2398795        1             2                   3            7   
2    473747        1             3                   3           12   
3   2254736        1             4                   4            7   
4    431534        1             5                   4           15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day     busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy   Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy  Least busy days   
2         Soda  ...  Mid-range product  Regularly busy  Least busy days   
3         Soda  ...  Mid-range product      Least busy  Least busy days   
4         Soda  ...  Mid-range product      Least busy  Least busy days   

  busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0        Average orders        10  New customer          6.37   Low spender   
1        Average orders        10  New customer          6.37   Low spender   
2           Most orders        10  New customer          6.37   Low spender   
3        Average orders        10  New customer          6.37   Low spender   
4           Most orders        10  New customer          6.37   Low spender   

  median_prior_orders   order_frequency_flag  
0                20.5  Non-frequent customer  
1                20.5  Non-frequent customer  
2                20.5  Non-frequent customer  
3                20.5  Non-frequent customer  
4                20.5  Non-frequent customer  

[5 rows x 25 columns]

In [58]:
df_orders_products_merged['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [59]:
print('Records in df_orders_products_merged with null values in column order_frequency_flag:')
df_orders_products_merged[df_orders_products_merged['order_frequency_flag'].isnull()][['user_id', 'days_since_prior_order', 'median_prior_orders', 'order_frequency_flag']]

Records in df_orders_products_merged with null values in column order_frequency_flag:


user_id  days_since_prior_order  median_prior_orders  \
13645692   159838                     NaN                  NaN   
17251990   159838                     NaN                  NaN   
17622767   159838                     NaN                  NaN   
24138593   159838                     NaN                  NaN   
25880002   159838                     NaN                  NaN   

         order_frequency_flag  
13645692                  NaN  
17251990                  NaN  
17622767                  NaN  
24138593                  NaN  
25880002                  NaN

In [60]:
df_orders_products_merged[df_orders_products_merged['user_id'] == 159838]

order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
13645692    895835   159838             1                   0           17   
17251990    895835   159838             1                   0           17   
17622767    895835   159838             1                   0           17   
24138593    895835   159838             1                   0           17   
25880002    895835   159838             1                   0           17   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
13645692                     NaN       10749                  3          0   
17251990                     NaN       33401                  6          0   
17622767                     NaN       23695                  2          0   
24138593                     NaN       21334                  5          0   
25880002                     NaN       22198                  1          0   

                                               product_name  ...  \
13645692                            Organic Red Bell Pepper  ...   
17251990                               Goat Cheese Crumbles  ...   
17622767                           California Veggie Burger  ...   
24138593                              Organic Peeled Garlic  ...   
25880002  4X Ultra Concentrated Natural Laundry Detergen...  ...   

            price_range_loc  busiest_day  busiest_days busiest_period_of_day  \
13645692  Mid-range product  Busiest day  Busiest days        Average orders   
17251990  Mid-range product  Busiest day  Busiest days        Average orders   
17622767  Low-range product  Busiest day  Busiest days        Average orders   
24138593  Mid-range product  Busiest day  Busiest days        Average orders   
25880002  Low-range product  Busiest day  Busiest days        Average orders   

         max_order  loyalty_flag average_price spending_flag  \
13645692         1  New customer          7.42   Low spender   
17251990         1  New customer          7.42   Low spender   
17622767         1  New customer          7.42   Low spender   
24138593         1  New customer          7.42   Low spender   
25880002         1  New customer          7.42   Low spender   

         median_prior_orders  order_frequency_flag  
13645692                 NaN                   NaN  
17251990                 NaN                   NaN  
17622767                 NaN                   NaN  
24138593                 NaN                   NaN  
25880002                 NaN                   NaN  

[5 rows x 25 columns]

In [61]:
df_orders_products_merged.shape

(32404859, 25)

In [62]:
# step 8
# feel like i can comment more on things. my brain goes straight to work and i have a feeling not commenting on things is going
# to come back and bite me eventually.

In [63]:
# step 9 export
df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_2.pkl'))